### ISSUE TITLE
[prim_count] clarification

### ISSUE THREAD

**Issue by cindychip** [lowRISC/opentitan#8343]
Hi Tim,

A few questions for prim_count when i am doing FPV testbench for it.

1).https://github.com/lowRISC/opentitan/blob/master/hw/ip/prim/rtl/prim_count.sv#L102
   Should this line be:
  `down_cnt <= max_val;`
  Because max_val is used to check cross counter sum, and it was gated by `OutSelDnCnt` but this line 102 did not.

2). We would like to clarify the spec as for how to use `set_cnt_i`.
     Looks like from the implementation:
     a). If counter is CrossCnt and OutSelDnCnt -> `set_cnt_i` will set the `max_val`.
     b). If counter is CrossCnt and OutSelUpCnt -> `set_cnt_i` won't set anything.
     c). If counter is DupCnt and OutSelUpCnt -> `set_cnt_i` will set the count up min value.
     d). If counter s DupCnt and OutSelDnCnt -> ASSERTION ERROR, down counter has to be cross count.
    Are the above summary correct?

@weicaiyang Thanks for debugging with me.

Thanks,
Cindy

**Comment by tjaychen** [comment]
let me add some more comments to this later to clarify.

> Hi Tim,
> 
> A few questions for prim_count when i am doing FPV testbench for it.
> 
> 1).https://github.com/lowRISC/opentitan/blob/master/hw/ip/prim/rtl/prim_count.sv#L102
> Should this line be:
> `down_cnt <= max_val;`
> Because max_val is used to check cross counter sum, and it was gated by `OutSelDnCnt` but this line 102 did not.

So this one is a bit odd... the `max_val` really only has utility for the `CrossCnt` case. 
So when we are doing cross cnt, the `max_val` and the `down_cnt` are set at the same time.  I think we can't set `down_cnt` to `max_val` because it will be offset by one cycle and we "might" get an error that way. 

This area is not super clean.. i need to think about how to make this better. 

> 
> 2). We would like to clarify the spec as for how to use `set_cnt_i`.
> Looks like from the implementation:
> a). If counter is CrossCnt and OutSelDnCnt -> `set_cnt_i` will set the `max_val`.

yes that's right, this ensure the up count does not count past the max_val and also latches the maximum value for us to compare.

> b). If counter is CrossCnt and OutSelUpCnt -> `set_cnt_i` won't set anything.

in this case it will use set_cnt_i to set the "starting" count of downcount.
But I think there's a bug here as max_val isn't set, so there's nothing to compare against. 

> c). If counter is DupCnt and OutSelUpCnt -> `set_cnt_i` will set the count up min value.

Yes that's right.

> d). If counter s DupCnt and OutSelDnCnt -> ASSERTION ERROR, down counter has to be cross count.
> Are the above summary correct?
> 
> @weicaiyang Thanks for debugging with me.
> 
> Thanks,
> Cindy

let me take a closer look at how to make this clearer. 

I think basically what I want is..

when doing `CrossCnt`, `set_cnt_i` should set the both the starting count of `down_cnt` and the maximum value.
when doing `DupCnt`, `set_cnt_i` should set the starting count of `up_cnt`.

**Comment by tjaychen** [comment]
does this #8345 help?

Also, i'm not sure if you guys want to wait on formally verifying this a bit.
When I added the prim count, I added it specifically for `keymgr` / `flash_ctrl`.  If we get everyone on board to use this for hardening, we most likely will need to add more features to it.

For example, I don't support cross count start of non-0 and non-max.

**Comment by cindychip** [comment]
Oh I see, I can wait until this is fully implemented :) Thanks for the heads up. Also I think this counter could be used for alert_handler. I believe there are some duplicated counters there.

**Comment by tjaychen** [comment]
yeah so i'm going to briefly go over this on monday. The ones to make sure
we match are alert_handler and aes (aes i think needs cross count, alert
handler needs dup cnt).
I'll need to check if other modules have something similar.  Kmac will
eventually need a version of this too I think.

**Comment by sriyerg** [comment]
IMO it will be great to upstream the secure design guidelines (from @cdgori). I cannot seem to find the link, so I am not sure if it is still under review. We probably should also update it clarifying which prim addresses which hardening feature. 

The design checklist (PR from Tom) already mandates the documentation & use of these prims. This will help get DV aligned in terms of how all secure hardening features are verified for each IP.

**Comment by tjaychen** [comment]
i believe @moidx is working through pushing it upstream. 
and yes, part of next week's meetings I hope we will go through which prims we have for what, what additional ones we should consider creating, and thinking about what else can be made common.

**Comment by tjaychen** [comment]
@cindychip can this issue be closed based on the clarifications so far?

**Comment by cindychip** [comment]
Yeah, thanks for the information.

**Comment by cindychip** [comment]
Hi Tim,

One more small question regarding prim_count.sv:
For `step_i` input, I believe the intention is that - once we set this value, it should not change unless we clear the counter.

If the assumption above is true, can we add some logic in design to ensure that step_i won't change, or ignore any changes of this input when counter is active?

Thanks,
Cindy

**Comment by cindychip** [comment]
Hi Tim,

Sorry I have one more question regarding cross counter to clarify:

Right now the code does not gate `set_i` to set new `set_cnt_i` when counter is enabled.
I believe the code works fine for the duplicated counter, but might have cause fatal alert in the cross counter.
This is the related code: https://github.com/lowRISC/opentitan/blob/master/hw/ip/prim/rtl/prim_count.sv#L64
   
Sequence:
    1). Set `set_cnt_i` to `100` and starts the counter with `en_i`, `step_i` = 1.
    2). Wait until `cnt_o` reaches `60`  -> right now `up_cnt_q` is 60 and `down_cnt` is 40
    3). Set `set_cnt_i` to `10` -> right now `up_cnt_q` = 61 and `down_cnt` = 10.  --> This will cause a fatal alert.
 
Could you help confirm if this sequence is legal and this behavior is intended?

Thanks again for your time and help.
Cindy

**Comment by cindychip** [comment]
https://github.com/lowRISC/opentitan/issues/8343#issuecomment-940624263
This is solved via : https://github.com/lowRISC/opentitan/pull/8624

**Comment by msfschaffner** [comment]
@cindychip can we remove the corresponding TODOs in the FPV config now?

https://github.com/lowRISC/opentitan/blob/b7daf29ceb4d78615c8434a0a1519f9213aaad9f/hw/top_earlgrey/formal/top_earlgrey_fpv_cfgs.hjson#L62-L89

**Comment by cindychip** [comment]
Sounds good, will create a PR soon.

**Comment by cindychip** [comment]
Done here： https://github.com/lowRISC/opentitan/pull/13596/files
